File: backend/app/models/simplefin.py
from pydantic import BaseModel, Field
from typing import Optional, List


class SetupTokenRequest(BaseModel):
    setup_token: str = Field(min_length=1)


class ImportTransactionRequest(BaseModel):
    transaction_id: int
    category_id: Optional[int] = None
    description: Optional[str] = None


class ImportAllRequest(BaseModel):
    account_id: Optional[int] = None
    default_category_id: Optional[int] = None


class ConnectionStatus(BaseModel):
    id: int
    institution_name: Optional[str] = None
    status: str
    last_synced: Optional[str] = None
    error_message: Optional[str] = None
    account_count: int = 0
    created_at: str
    updated_at: str


class LinkedAccountResponse(BaseModel):
    id: int
    connection_id: int
    simplefin_account_id: str
    name: str
    institution: Optional[str] = None
    balance_cents: Optional[int] = None
    available_balance_cents: Optional[int] = None
    currency: str = "USD"
    account_type: Optional[str] = None
    last_synced: Optional[str] = None


class SyncedTransactionResponse(BaseModel):
    id: int
    account_id: int
    simplefin_transaction_id: str
    amount_cents: int
    description: str
    posted_date: Optional[str] = None
    transacted_date: Optional[str] = None
    pending: bool = False
    category: Optional[str] = None
    imported: bool = False
    imported_transaction_id: Optional[int] = None
    created_at: str
    account_name: Optional[str] = None
    account_institution: Optional[str] = None


class SyncResult(BaseModel):
    new_accounts: int = 0
    updated_accounts: int = 0
    new_transactions: int = 0
    updated_transactions: int = 0


class PaginatedSyncedTransactions(BaseModel):
    transactions: List[SyncedTransactionResponse]
    total: int
    page: int
    per_page: int
    total_pages: int

File: frontend/src/types/simplefin.ts
export interface SimpleFinConnection {
  id: number;
  institution_name: string | null;
  status: 'active' | 'error' | 'disconnected';
  last_synced: string | null;
  error_message: string | null;
  account_count: number;
  created_at: string;
  updated_at: string;
}

export interface LinkedAccount {
  id: number;
  connection_id: number;
  simplefin_account_id: string;
  name: string;
  institution: string | null;
  balance_cents: number | null;
  available_balance_cents: number | null;
  currency: string;
  account_type: string | null;
  last_synced: string | null;
}

export interface SyncedTransaction {
  id: number;
  account_id: number;
  simplefin_transaction_id: string;
  amount_cents: number;
  description: string;
  posted_date: string | null;
  transacted_date: string | null;
  pending: boolean;
  category: string | null;
  imported: boolean;
  imported_transaction_id: number | null;
  created_at: string;
  account_name: string | null;
  account_institution: string | null;
}

export interface SyncResult {
  new_accounts: number;
  updated_accounts: number;
  new_transactions: number;
  updated_transactions: number;
}

export interface PaginatedSyncedTransactions {
  transactions: SyncedTransaction[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}
